From 6882a1bb78c8ce449e721dac25452a4a1d94ad98 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Sat, 03 Sep 2022 10:35:22 +0300
Subject: [PATCH] [SW-0] gaudi3: update axprot for debug if

Currently axprot values for debug if is only set during init in case
the efuse bit is set to "disable debug if" as debug ifs are privileged
by default. But as these efuse bits can be also modified using fake
fuse, there is a possibility to switch the debug if values from
DIS -> EN in hard/fw resets also.

Hence the logic in rom also needs to make the debug ifs back to
privilege in such cases.

Change-Id: Id9d199ac8cc2464a2182318e917f82ed8aa08c13
---

diff --git a/arch/gaudi3/arc/dev_security.h b/arch/gaudi3/arc/dev_security.h
index 20aba1c..2af93ff 100644
--- a/arch/gaudi3/arc/dev_security.h
+++ b/arch/gaudi3/arc/dev_security.h
@@ -74,6 +74,14 @@
 	RR_MAX_CTX_TYPES,
 };
 
+/*
+ * Enum to set axprot to privilege or secured level in rom
+ */
+enum rom_axprot_level_config {
+	AXPROT_PRIV,
+	AXPROT_SEC,
+};
+
 int secureboot_verify_image_signature(const uint8_t *image, uint32_t image_len, uint8_t *key,
 				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
 				uint8_t sec_ver, uint8_t key_index);
diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
index e4c2a4b..3c899e0 100644
--- a/arch/gaudi3/arc/security.c
+++ b/arch/gaudi3/arc/security.c
@@ -36,6 +36,7 @@
 #define AXPROT_OVRD_VAL_POS			4
 #define AXPROT_SEC_VAL		((PROT_OVR_SECURE << AXPROT_OVRD_VAL_POS) | AXPROT_OVRD_EN)
 #define AXPROT_UNPRIV_VAL	((PROT_OVR_UNPRIV << AXPROT_OVRD_VAL_POS) | AXPROT_OVRD_EN)
+#define AXPROT_PRIV_VAL		((PROT_OVR_PRIV << AXPROT_OVRD_VAL_POS) | AXPROT_OVRD_EN)
 #define RR_DEFAULT_VAL				0
 /*
  * This key hash is calculated on the Public Key Properties -
@@ -490,26 +491,36 @@
 void security_smi_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 void security_jtag_rr_hbw_config(__unused uint8_t spi_controller_en) {}
 
-static void security_i2c_axprot_config(void)
+static void security_i2c_axprot_config(enum rom_axprot_level_config prot_lvl)
 {
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	uint32_t axprot_priv_reg_val = AXPROT_PRIV_VAL;
+
+	if (prot_lvl == AXPROT_SEC)
+		axprot_priv_reg_val = AXPROT_UNPRIV_VAL;
+
+	regmap_writel(axprot_priv_reg_val, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
 }
 
-static void security_smi_axprot_config(void)
+static void security_smi_axprot_config(enum rom_axprot_level_config prot_lvl)
 {
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	uint32_t axprot_priv_reg_val = AXPROT_PRIV_VAL;
+
+	if (prot_lvl == AXPROT_SEC)
+		axprot_priv_reg_val = AXPROT_UNPRIV_VAL;
+
+	regmap_writel(axprot_priv_reg_val, mmD0_SMI_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_SMI_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_SMI_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
 }
@@ -531,6 +542,8 @@
 	 * axprot.
 	 */
 	security_smi_access(SMI_EN_VAL);
+	security_i2c_axprot_config(AXPROT_PRIV);
+	security_smi_axprot_config(AXPROT_PRIV);
 }
 
 int security_efuse_hw_rr_config(void)
@@ -544,8 +557,8 @@
 void security_disable_twi_if(void)
 {
 	security_smi_access(SMI_DIS_VAL);
-	security_i2c_axprot_config();
-	security_smi_axprot_config();
+	security_i2c_axprot_config(AXPROT_SEC);
+	security_smi_axprot_config(AXPROT_SEC);
 
 	if (!security_efuse_hw_rr_config()) {
 		security_rr_twi_lbw_config(&rr_ctx_lbw[RR_CLOSE_ALL]);
@@ -567,14 +580,19 @@
 	regmap_writel(status, mmD0_PSOC_CS_TRACE_SECURITY_CTRL);
 }
 
-static void security_jtag_axprot_config(void)
+static void security_jtag_axprot_config(enum rom_axprot_level_config prot_lvl)
 {
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	uint32_t axprot_priv_reg_val = AXPROT_PRIV_VAL;
+
+	if (prot_lvl == AXPROT_SEC)
+		axprot_priv_reg_val = AXPROT_UNPRIV_VAL;
+
+	regmap_writel(axprot_priv_reg_val, mmD0_JT_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_JT_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
-	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_JT_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(axprot_priv_reg_val, mmD0_JT_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
 	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
 }
@@ -582,7 +600,7 @@
 void security_disable_jtag_if(void)
 {
 	security_jtag_dap(SECURITY_JTAG_DAP_DISABLE);
-	security_jtag_axprot_config();
+	security_jtag_axprot_config(AXPROT_SEC);
 
 	if (!security_efuse_hw_rr_config()) {
 		security_rr_jtag_lbw_config(&rr_ctx_lbw[RR_CLOSE_ALL]);
@@ -593,6 +611,7 @@
 void security_enable_jtag_if(void)
 {
 	security_jtag_dap(SECURITY_JTAG_DAP_ENABLE);
+	security_jtag_axprot_config(AXPROT_PRIV);
 }
 
 void security_handle_jtag_if(void)
